ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -15,52 +15,258 @@
 
 package io.confluent.ksql.internal;
 
-import io.confluent.ksql.engine.QueryEventListener;
-import io.confluent.ksql.metastore.MetaStore;
-import io.confluent.ksql.services.ServiceContext;
-import io.confluent.ksql.util.QueryMetadata;
+import com.google.common.collect.ImmutableMap;
+import io.confluent.ksql.util.KsqlConfig;
+
+import java.io.File;
+import java.math.BigInteger;
 import java.util.ArrayList;
-import java.util.LinkedList;
+import java.util.Collection;
+import java.util.HashMap;
 import java.util.List;
-import org.apache.kafka.common.utils.Time;
-import org.apache.kafka.streams.KafkaStreams;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.common.metrics.Gauge;
+import org.apache.kafka.common.metrics.KafkaMetric;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.MetricsContext;
+import org.apache.kafka.common.metrics.MetricsReporter;
+import org.apache.kafka.streams.StreamsConfig;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class UtilizationMetricsListener implements Runnable, QueryEventListener {
+public class UtilizationMetricsListener implements MetricsReporter {
+  private final ConcurrentHashMap<String, Long> queryStorageMetrics;
+  private final Map<String, List<String>> metrics;
+  private final Metrics metricRegistry;
+  private final File baseDir;
+  private final Collection<TaskStorageMetric> registeredMetrics;
+
+  public UtilizationMetricsListener(final KsqlConfig config, final Metrics metricRegistry) {
+    this.queryStorageMetrics = new ConcurrentHashMap<>();
```

Comment:
Could you elaborate a bit more about the differences of these three collections: the `metrics` map, the `queryStorageMetrics` map, and the `registeredMetrics` set. From what I can see it seems:

* `queryStorageMetrics` is used for the per-query metrics, but it was not populated and updated anywhere.
* `registeredMetrics` is used for per-query-task metrics
* `metrics` is also used for storing per-query metrics?

The relationships of these and how they are leveraged / updated are all not very clear to me.